### PR TITLE
POC DO NOT MERGE - hide inference for semantic text 

### PR DESCRIPTION
**Pull request by carlosdelest**
A POC showing how we could remove inference fields for semantic_text results.

It could be simpler if we just hide a field pattern - but "inference" seems too common a word. In case we use something like "_semantic_text_inference" for storing inference results, we might simplify this as we could avoid obtaining the inference fields to remove them from the `FetchSourceContext`.

This POC:
- Adds a boolean field to `SearchRequestBuilder` to hide source fields (so we can tell update by query / reindex / other queries apart)
- Adds a factory method for `FetchSourceContext` to add the inference fields to the list of excluded fields if needed
- Modifies `TransportGetAction` and `SearchService` to create a new `FetchSourceContext` that takes into account inference fields

Does this sound like a good approach for hiding inference results?

### How this works

Inference results are hidden by default:

`GET test-sparse/_doc/doc1` will hide results

`GET test-sparse/_doc/doc1?_source_includes=*` will display them.

The following hides results:
```json
GET test-sparse/_search
{
    "query": {
        "semantic": {
            "field": "infer_field",
            "query": "burger"            
        }
    }
}
```

The following displays them:

```json
GET test-sparse/_search?_source_includes=*
{
    "query": {
        "semantic": {
            "field": "infer_field",
            "query": "burger"            
        }
    }
}


```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -203,6 +203,8 @@ public static HighlightBuilder highlight() {
 
     private Map<String, Object> runtimeMappings = emptyMap();
 
+    private boolean hideSourceFields = true;
```
Comment (Mikep86) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:206]: Can we change the name of the flag to be more descriptive? Like `hideInferenceResultFields`?

**Review by Mikep86 — COMMENTED** [review]
I like the approach! Simple and effective :)

**Review by jimczi — COMMENTED** [review]
The approach makes sense but I wonder if we can make it more generic. 
I am thinking of something like:
```
"_source": {
  "include_vectors": false
}
```
This way this is only a boiler plate to exclude the fields that contain vector (dense or sparse). It is also part of the `_source` parameter since that's the context where the option applies. 
I am also not sure whether the default should be to not include the vectors. The goal for this option is to remove vectors from the response without having to know all the fields to exclude in advance. 
So my preference would be to continue to return vectors by default and expose this option to opt-out easily.

**Comment by carlosdelest** [comment]
@jimczi I'm fine with that approach. The only problem I see is that we wanted to [hide ELSER tokens by default](https://elasticco.atlassian.net/issues/ML-405?jql=text%20~%20%22hide%20elser%22), this was one of the design targets for `semantic_text`.

Maybe our users are already educated on this and we don't need to hide the tokens anymore, so we can provide this as an alternative.

What are your thoughts?

**Comment by jimczi** [comment]
I have a slight preference to not hide things magically. I'd rather rephrase the requirement as "we need a simple way to disable returning dense/sparse vectors from search response". Currently it is contextual and you need to know the field names for exclusion. With an explicit parameter that would be straightforward and applications could decide which default they want to apply on this feature.

**Pull request closed without merging**